### PR TITLE
Added lstm_layer class

### PR DESCRIPTION
**Pull request by n-Guard**

### CODE REVIEW

**Comment by n-Guard** [comment]
Allright the tests seem to be working but there are some caveats: 
Right now testing only works because the Keras input to its LSTM layer has just one timestep. If it would have more timesteps, the output generated by Keras and fdeep output would not be the same (it applies only to the tests though). This is due to the fact that the function that generates dummy test data does not know the number of timesteps, as it is our `tensor3s.size()` and can't easily be specified inside shape3. 
For the moment I left the swapaxes stuff as it is, so the number of timesteps is actually still inside the tensor3 element as `width` but is ignored to not break anything else.

**Comment by Dobiasd** [comment]
Yeah! This is very nice!

Do you think it makes sense to add further test cases? E.g.:
- with LSTM after another layer type
- with LSTM in front of another layer type
- with different input shapes?
- with different activation functions?

**Comment by n-Guard** [comment]
Yes, I will soon add some additional test cases to be absolutely sure that everything works!

How would you address the issue that `seq_len` is now outside of shape3 and therefore cannot be integrated into tests at the moment?

**Comment by Dobiasd** [comment]
Why can `seq_len` not be integrated into the tests because of this?
We [also test layer types like `Concatenate` by feeding them multiple tensor3s](https://github.com/Dobiasd/frugally-deep/blob/939dce48bbd32952740ea1523b21331ba2d45f85/keras_export/generate_test_models.py#L86). Is there a reason this approach would not work with LSTM?

**Comment by Dobiasd** [comment]
Mhh, OK. So maybe then we should not feed the LSTM layer in C++ with `(seq_len, n_features, 1, 1)`, but instead use the equivalent to what is used in Keras (swapped accordingly). Is this possible?

**Comment by n-Guard** [comment]
So just that I understand correctly: Do you propose to drop our agreement on using 4-dim input (that is `tensor3s`) and use 3-dim input instead? 
If so, then the question would be: 

Leave everything as it is (with swapaxes etc.) and use (`n_features, 1, seq_len`) as input, which is a little bit confusing to be honest, but would work.
**OR:**
Use (`seq_len, n_features, 1`) as input which is imho the most consistent solution and closest to "what Keras does". 

In order to make the second solution work, fdeep format has to be changed to "channels last". As I recall you were open to change it. How much work do you think would this be?

**Comment by Dobiasd** [comment]
> So just that I understand correctly: Do you propose to drop our agreement on using 4-dim input (that is tensor3s) and use 3-dim input instead?

I'd like to have it simple and clean. And if this is the only way to achieve it, yes, why not.

> In order to make the second solution work, fdeep format has to be changed to "channels last". As I recall you were open to change it. How much work do you think would this be?

In case everything works out perfectly from the beginning, I would estimate this change to take a maximum of 4 hours of work. So in reality it's probably more like 2 days. ;)

If we agree on this as being the best solution, I'd try to do it on the weekend.

**Comment by n-Guard** [comment]
> Since the historical choice of CHW over HWC more and more turned out to be technical dept, I just made the needed changed. Yes, that breaks the API, but makes the code a lot simpler overall. ☺️

Wow, this is really great!! ❤️ 🎉
I didn't expect it to be this soon! It definitely makes things simpler.

> (It turned out to only be about 8 hours of work. The most difficult part was to get im2col correct and fast again.)

I'm glad this turned out to be a somewhat reasonable amount of work (but still work though 😉).

While I tried to apply the changes to my lstm_layer class, I noticed that the input format now would be `(1, seq_len, n_features)` which differs from my proposed `(seq_len, n_features, 1)` but if you think it's more consistent to export it this way, then we can leave it like that since it definitely makes more sense than `(n_features, 1, seq_len)`.

**Comment by Dobiasd** [comment]
I think we should use whatever Keras uses, to avoid confusion. :)

**Comment by n-Guard** [comment]
That's true but we still need to translate a 2-dim or 1-dim Keras shape into an fdeep shape3 (or shape_hwc) so we have to add new dimensions which are completely independent of Keras. 
I guess this is a matter of personal taste now if we "fill" tensor3 from right to left or vice versa, so we can go with your version :)

**Comment by Dobiasd** [comment]
OK, so for images we strictly map Keras' `(batch_size==1, height, width, channels)` to fdeep's `(height, width, channels)`, which makes sense, but for other things like `(batch_size==1, sequence_length)` or `(batch_size==1, sequence_length, output_dim)`, we just have to choose which of `height`, `width`, `channels` we want to "misuse".

So, what would make more sense? Left-padding like `(1, 1, sequence_length)` and `(1, sequence_length, output_dim)`, or right-padding like `(sequence_length, 1, 1)` and `(sequence_length, output_dim, 1)`. (in case of [`keras.layers.Embedding`](https://keras.io/layers/embeddings/) as an example)

My first intuition was to prefer the left-padded version, but the more I think about it, the right-padded version looks better, since it does have the advantage of `sequence_length` being in the same position in both examples. So this would point to the `(seq_len, n_features, 1)` version, you already proposed.

**Comment by n-Guard** [comment]
> My first intuition was to prefer the left-padded version, but the more I think about it, the right-padded version looks better, since it does have the advantage of sequence_length being in the same position in both examples. So this would point to the (seq_len, n_features, 1) version, you already proposed.

You are right, I also would say that right-padding would make more sense so dimensions can be kept in place as in your example.

**Comment by n-Guard** [comment]
Allright, no problem! 
There are still two things for your consideration:
1. Right now when a user inputs a `tensor3s` with `size = 1` into the `lstm_layer` we have our `(seq_len, n_features, 1)` as we agreed upon, which is congruent with Keras. 
But I added a feature that if `tensor3s.size() > 1`, `tensor3s.size()` is used as time dimension instead and `tensor3.shape()` must be `(n_features, 1, 1)`. I know this is somewhat against what we discussed earlier (not keeping dims in place), but it would make things easier when implementing `TimeDistributed` layer. As an example, when we have `TimeDistributed(Flatten())` layer which is a common layer before a recurrent layer, we could utilize `tensor3s.size()` as time dimension again in order to leave `Flatten()` as it is. 

2. When using larger input dims (especially for seq_len) the values between Keras and fdeep sometimes start to differ slightly more than 10^-5. I'm guessing this is due to the recursion and I'm not sure if this is a problem.

**Comment by Dobiasd** [comment]
1. I'd have no problem with modifying `flatten`. Currently [it does not even accept more than one tensor](https://github.com/Dobiasd/frugally-deep/blob/master/include/fdeep/layers/flatten_layer.hpp#L27), so we could add something for that case. I think I currently have no string opinion into either direction. Choose what you think is better in the long run. :+1:

2. Have you already tried with `#define FDEEP_FLOAT_TYPE double`?

**Comment by n-Guard** [comment]
> I'd have no problem with modifying `flatten`. Currently it does not even accept more than one tensor, so we could add something for that case. I think I currently have no string opinion into either direction. Choose what you think is better in the long run. 👍

I guess we have to think a little bit more general than just modifying `flatten`. Correct me if I'm wrong but I think we already have the `TimeDistributed` functionality in fdeep, i.e. using `model.predict()` with a `tensor3s` (e.g. with `size > 1`). Every `tensor3s` element is fed separately to the model and produces a separate output, right? This is exactly what we want when using `TimeDistributed` wrapper, so we would have our time series as `tensor3s` elements. I'd like to utilize this fact, so that an lstm_layer can understand these separate predictions (aka elements in `tensor3s`). This way we could use every non-recurrent layer with `TimeDistributed` wrapper. But I'm not quite sure yet how to implement it best.

> Have you already tried with `#define FDEEP_FLOAT_TYPE double`?

Thanks for the suggestion, with double precision even larger inputs indeed work fine :)

**Comment by Dobiasd** [comment]
> Correct me if I'm wrong but I think we already have the TimeDistributed functionality in fdeep, i.e. using model.predict() with a tensor3s (e.g. with size > 1). Every tensor3s element is fed separately to the model and produces a separate output, right?

No, this is not what it is doing. In Keras a model (at least created with the functional API) can have multiple input tensors and output tensors. Multiple input tensors can not be fed separately. Example:

We create a model that takes two sensors as input and outputs two, but they are not processed in isolation, but combined.
```python
from keras.models import Model
from keras.layers import Input, Concatenate, Add

inputs = [
    Input(shape=(240, 320, 3)),
    Input(shape=(240, 320, 3))
]

outputs = [
    Concatenate()([inputs[0], inputs[1]]),
    Add()([inputs[0], inputs[1]])
]

model = Model(inputs=inputs, outputs=outputs)
model.compile(loss='mse', optimizer='nadam')
model.save('multi_input_and_output_model.h5', include_optimizer=False)
```

Now in C++, we need to also provide two tensors:

```cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("multi_input_and_output_model.json");
    const auto result = model.predict({
        fdeep::tensor3(fdeep::shape_hwc(240, 320, 3), 42),
        fdeep::tensor3(fdeep::shape_hwc(240, 320, 3), 43)
        });
    std::cout << fdeep::show_tensor3s(result) << std::endl;
}
```

Does this make sense?

**Comment by n-Guard** [comment]
Oh right, that definitely makes sense, thanks for the explanation! 👍

I guess then it's also not possible to call `model.predict() `in a loop inside `TimeDistributed` for every `tensor3s` element when our model only has one input, right? Because it gets problematic when we have a model with multiple inputs and outputs that also has `TimeDistributed` somewhere in it. 
Ok, the only solution I see at the moment is a `shape4` to get `TimeDistributed` working properly. Or do you have another idea?

**Comment by Dobiasd** [comment]
Yeah, `model.predict()` currently is not intended to store any state between calls. So calling it in a loop might not be a good option.

So for `TimeDistributed` to work, we would need to switch from `tensor3` to `tensor4` and from `shape_hwc` to something like `shape_thwc`, right?

Do you need `TimeDistributed` for what you are actually doing with frugally-deep, or would it be just a nice optional thing to have?

File: include/fdeep/recurrent_ops.hpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+#pragma once
+
+#include <string>
+#include <functional>
+
+namespace fdeep { namespace internal
+{
+
+static float_type linear_activation(float_type x)
```
Comment (Dobiasd) [inline include/fdeep/recurrent_ops.hpp]: As far as I know the `static` in front of free functions means two things:
- Semantics: It allows other translation units to have different functions with the same name.
- Performance optimizations: It allows inlining into other translation units.

Since frugally-deep is a header-only library, I think both points are irrelevant in our case, and I see no point of using `static` here. Please correct me if I'm missing something. Otherwise please remove the `static`. :)

File: include/fdeep/recurrent_ops.hpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+{
+    return x;
+}
+
+static float_type tanh_activation(float_type x)
+{
+    return std::tanh(x);
+}
+
+static float_type sigmoid_activation(float_type x)
```
Comment (Dobiasd) [inline include/fdeep/recurrent_ops.hpp]: Are these ops the same ops as the non-recursive ops currently coded in the layers? If so, we might unify them to avoid code duplication.

Comment (Dobiasd) [reply include/fdeep/recurrent_ops.hpp]: But maybe we can do this in a separate clean-up PR after the new functionality here is merged.

**Comment by n-Guard** [comment]
> So for `TimeDistributed` to work, we would need to switch from `tensor3` to `tensor4` and from `shape_hwc` to something like `shape_thwc`, right?

Yes, if we want to use it with all fdeep layers (especially `Conv2D`) we would need something like a `shape_thwc` then.

> Do you need `TimeDistributed` for what you are actually doing with frugally-deep, or would it be just a nice optional thing to have?

Unfortunately, I need the `TimeDistributed` layer for what I'm doing, so I have to implement it somehow. At the moment I don't need all four dimensions though because I only have `Conv1D` wrapped with `TimeDistributed`, so we could limit the functionality to 3 dimensions for now. We then could have 2-dim feature input that is wrapped by `TimeDistributed` resulting in `(time, feature_dim_1, feature_dim_2)`.

File: include/fdeep/recurrent_ops.hpp (hunk 1 above)
Comment (n-Guard) [reply include/fdeep/recurrent_ops.hpp]: Yes you are right, I will remove `static`!

File: include/fdeep/recurrent_ops.hpp (hunk 2 above)
Comment (n-Guard) [reply include/fdeep/recurrent_ops.hpp]: Those are exactly the same, so unifying them would definitely make sense.

**Comment by Dobiasd** [comment]
You can add `Bidirectional` here, if you like. But for me the end product is more important than the commit history. :)

---

Regarding `TimeDistributed` we can now choose between

- breaking the public API again by switching to 4d tensors,

- or implementing a hack that makes it work with 3d tensors for a subset of all use cases, but that might be technical dept in the long run.

Right?

---

How would we design the tensor/shape classes if we would be writing fdeep from scratch to avoid headaches like the current one?

**Comment by n-Guard** [comment]
> Regarding `TimeDistributed` we can now choose between
> 
>     * breaking the public API again by switching to 4d tensors,
> 
>     * or implementing a hack that makes it work with 3d tensors for a subset of all use cases, but that might be technical dept in the long run.
> 
> 
> Right?

Yes, I would say these are the options.

> How would we design the tensor/shape classes if we would be writing fdeep from scratch to avoid headaches like the current one?

Maybe it's worth to have a look at Eigen-unsupported again, since it's also used in Tensorflow to construct the tensors. In theory this would allow us to have n-dimensional tensors so we don't need to rely on a fixed shape. There are also many handy matrix operations already implemented (e.g. easy reshaping of dimensions).

**Comment by Dobiasd** [comment]
If at some point in the future we don't need Eigen for fast GEMM any more, I'd happily drop this dependency, thus I'd like to avoid becoming tightly coupled to it.

---

So if we would switch fdeep to higher-dimensional tensors, would four be enough, or would we perhaps need more later? Is there a maximum number of dimensions in Keras?

In case four would be enough, maybe we should name them more content agnostic, like `shape4` and `tensor4::set` etc.

**Comment by n-Guard** [comment]
> So if we would switch fdeep to higher-dimensional tensors, would four be enough, or would we perhaps need more later? Is there a maximum number of dimensions in Keras?

Technically there isn't a maximum since `Dense` accepts n-dimensional inputs, but on the other hand this doesn't play such a big role because the input can always be flattened without problems. 
I think 5 dimensions would be enough for most of the layers. If we had a `shape5` there would be room for `Conv3D` (4-dimensional) and even `TimeDistributed(Conv3D)` (5-dimensional).

---

Apart from the input dimension problem, I'm wondering what's the best way to implement `TimeDistributed`. Basically we need some kind of functionality that allows us to call a layer (e.g. `Conv2D`) inside another layer (i.e. `TimeDistributed`) for every timestep, aggregate results for all timesteps and pass it to the next layer. Do you have any idea what the best approach would be?

**Comment by Dobiasd** [comment]
OK, I'm online again, and will try to do it during the next few days.

**Comment by Dobiasd** [comment]
OK, just pushed [the large commit](https://github.com/Dobiasd/frugally-deep/commit/1b1a706b4a95eda9e2d0dce6294c937428142db0). :)

We now have `shape5` and `tensor5`.

I "right-aligned" (i.e., left-padded) them, so currently it is
`shape5(size_dim_5, size_dim_4, height, width, depth)`
and
`tensor5_pos(pos_dim_5, pos_dim_4, y, x, z)`.

Do you think this is OK?

---

> I think 5 dimensions would be enough for most of the layers.

This reminds me of "640 k ought to be enough for anybody." :D

I'll quote you on that. ;)

**Comment by Dobiasd** [comment]
> So for the LSTM layer the shape would be (1, 1, 1, seq_len, n_features) right?

Yes, if it is `(seq_len, n_features)` in Keras, then it should be `(1, 1, 1, seq_len, n_features)` in frugally-deep.

---

> I got the `TimeDistributed` functionality working, should I also add it to this PR?

Yes, sounds OK to me.

**Comment by Dobiasd** [comment]
How is it going? Any blockers? Let me know when you think this PR is done and the features have good test coverage, so then I'll review. :)

**Comment by n-Guard** [comment]
I think you can review, please let me know if you think there should be more testing cases!
I only tested `TimeDistributed` with up to 4 dimensions because for some reason the Keras functional API didn't let me specify an input greater than 4. With a sequential model this is no problem, but in order to use it with our existing `test_model_sequential` I would have needed `reshape_layer` to support 5 dimensions which isn't implemented in fdeep yet. `time_distributed_layer` in fdeep can handle up to 5 dimensions though.

File: include/fdeep/import_model.hpp [hunk 3]
```diff
@@ -839,6 +842,85 @@ inline nodes create_nodes(const nlohmann::json& data)
...
+    const std::string unit_activation = data["config"]["activation"];
+    const std::string recurrent_activation = data["config"]["recurrent_activation"];
+    const bool use_bias = data["config"]["use_bias"];
+
+    RowMajorMatrixXf bias(1, units * 4);
+    if (use_bias)
+        bias = eigen_row_major_mat_from_values(1, units * 4, decode_floats(get_param(name, "bias")));
+
+    const float_vec W = decode_floats(get_param(name, "weights"));
+    const RowMajorMatrixXf weights = eigen_row_major_mat_from_values(W.size() / (units * 4), units * 4, W);
```
[the remark this answers is not in the document]
Comment (n-Guard) [reply include/fdeep/import_model.hpp]: Yes, that definitely makes sense, I will change it!

File: keras_export/convert_model.py [hunk 4]
```diff
@@ -354,8 +443,14 @@ def get_all_weights(model):
             assert is_ascii(name)
             if name in result:
                 raise ValueError('duplicate layer name ' + name)
-            if show_func:
+            if show_func and show_func(layer) != None:
```
[the remark this answers is not in the document]
Comment (n-Guard) [reply keras_export/convert_model.py]: Not yet, but I will add that :)

File: keras_export/convert_model.py [hunk 5]
```diff
@@ -303,6 +303,92 @@ def show_prelu_layer(layer):
...
+
+    if len(forward_weights) == 3:
+        result['forward_bias'] = encode_floats(forward_weights[2])
+    if len(backward_weights) == 3:
+        result['backward_bias'] = encode_floats(backward_weights[2])
+
+    return result
+
+def show_time_distributed_layer(layer):
+    show_layer_functions = {
```
[the remark this answers is not in the document]
Comment (n-Guard) [reply keras_export/convert_model.py]: I made a separate function that returns this dict so now both functions can use it.

**Comment by n-Guard** [comment]
Ok so now everything except the activation functions duplication is as you suggested :)

**Comment by Dobiasd** [comment]
:+1: :100: :1st_place_medal: :tada:

**Comment by Dobiasd** [comment]
Thank you very much again. :heart:

This has easily been the largest PR to this library so far. And it not only extended the functionality by some important things, but the accompanied discussions also yielded general changes in the lib, that made it more future proof. :+1:

**Comment by n-Guard** [comment]
Thank you very much for your kind words, I'm glad I could help extend this library! :)
I appreciate your support and dedication to this project and I'm sure this will help many people realise their ideas.
